airflow/decorators/__init__.py

```diff
@@ -41,103 +46,13 @@ def __call__(
-            defined with def, which will be run in a virtualenv
-        :type python_callable: function
-        :param multiple_outputs: if set, function return value will be
-            unrolled to multiple XCom values. List/Tuples will unroll to xcom values
-            with index as key. Dict will unroll to xcom values with keys as XCom keys.
-            Defaults to False.
-        :type multiple_outputs: bool
-        :param requirements: A list of requirements as specified in a pip install command
-        :type requirements: list[str]
-        :param python_version: The Python version to run the virtualenv with. Note that
-            both 2 and 2.7 are acceptable forms.
-        :type python_version: Optional[Union[str, int, float]]
-        :param use_dill: Whether to use dill to serialize
-            the args and result (pickle is default). This allow more complex types
-            but requires you to include dill in your requirements.
-        :type use_dill: bool
-        :param system_site_packages: Whether to include
-            system_site_packages in your virtualenv.
-            See virtualenv documentation for more information.
-        :type system_site_packages: bool
-        :param op_args: A list of positional arguments to pass to python_callable.
-        :type op_args: list
-        :param op_kwargs: A dict of keyword arguments to pass to python_callable.
-        :type op_kwargs: dict
-        :param string_args: Strings that are present in the global var virtualenv_string_args,
-            available to python_callable at runtime as a list[str]. Note that args are split
-            by newline.
-        :type string_args: list[str]
-        :param templates_dict: a dictionary where the values are templates that
-            will get templated by the Airflow engine sometime between
-            ``__init__`` and ``execute`` takes place and are made available
-            in your callable's context after the template has been applied
-        :type templates_dict: dict of str
-        :param templates_exts: a list of file extensions to resolve while
-            processing templated fields, for examples ``['.sql', '.hql']``
-        :type templates_exts: list[str]
-        """
-        return _virtualenv_task(
-            python_callable=python_callable,
-            multiple_outputs=multiple_outputs,
-            requirements=requirements,
-            python_version=python_version,
-            use_dill=use_dill,
-            system_site_packages=system_site_packages,
-            string_args=string_args,
-            templates_dict=templates_dict,
-            templates_exts=templates_exts,
-            **kwargs,
-        )
+    def __getattr__(self, name):
+        if self.store.get(name, None):
+            return self.store[name]
+        connections = [e for e in metadata.entry_points()['airflow.task_decorators'] if e.name == name]
```

Comment:
Why do we want to make a separate entrypoint rather than make it an entry in provider_info one ?  Seems like unnecessary duplication of entrypoints. We already have provider_info schema where we validate the dictionary passed there and it is fully extendable (and makes it one place to keep all details about the provider).

Reply:
Primarily because you then don't have to load _all_ plugins and pay the import time cost just to use one.

(There is no need to validate anything here -- it's a single name and a function)

Reply:
But we do it anyway? The commit mentions that this is all about providers, so the providers will be loaded anyway at that point I believe? Or if we foresee that as not-part-of-providers, this should be a separate non-provider-related feature.

Reply:
The entrypoint metadata is loaded, but iterating over entrypoints on this line doesn't load any python code.

Reply:
> As of Airflow 2.3, users can now integrate custom decorators into their provider packages and have those decorators

You can easily add providers without any hooks/operator custom connection extensions and you will not pay the price of importing, so I think it does not change anything when it comes to performance

Reply:
> The entrypoint metadata is loaded, but iterating over entrypoints on this line doesn't load any python code.

Neither provider.yaml if it has no cusom hooks/connections/operators defined.

Reply:
I really think it's just unnecessary duplication of entrypoints - which might bring a lot of confusion and brings no benefits (unless there are some other reasons I am not aware of of course) - but if make it part of "provider" capabilities, and some other separate "feature", I see no reason why we should split that off.

Reply:
Also we have the docs of "How to write a custom provider" - https://airflow.apache.org/docs/apache-airflow-providers/index.html#custom-provider-packages - if you try to fit that in this documentation (which we should if we make it part of provider's capabilities) you will see yourself how big of a confusion it might introduce.

Reply:
That actually reminded me of something that I wanted to do for some time. I thought that we can further optimise loading of provider hooks and extra links (thus imports)  by splitting the initialize_providers_manager() method and only running what is needed when we access properties. 

The main effect is speeding up some of the command lines : https://github.com/apache/airflow/pull/17304

But there are scenarios where we shave-off couple of seconds, when for example Tasks are never accessing .get_conn() method (that one requires to import all hooks)  but for some reasons would like to access list of Providers (those this scenario is quire unlikely). 

However if opens up the same pattern for task_decorators - you can add lazy task_decorator initialization in the same way and then workers that never access get_conn() but they need decorator, will not needlessly import the connections.

Reply:
I think with the #17304, there is no scenario, where having connection classes, operator extra link classes (and task_decorators)  defined in provider_info, would negatively impact performance. They are truly lazy loaded when needed

Reply:
RIght now the plugin manager has to load everything (of a type) when it loads anything for that type -- i.e. asking for one hook/connection loads _all_ hooks. - that is the primary difference to this approach here -- we can load a single thing by name.

Looking at your change it does address most of my issues with the current approach, and while we do still load all provider infos at once, that's not _too_ bad (so long as we warm the cache in the DagFileProcessorManager so that each parser process doesn't need to load it each time it parses a dag file -- which this PR currently also doesn't get right)

@dimberman Can you change this PR to use a new field in the existing provider_info.yaml approach. Using that we can still have the "only load the decorator code on first access" approach, but all that should live in plugin_manager and use.

Something like this as the (new) fields in provider info:

```
---
package-name: apache-airflow-providers-docker
name: Docker
description: |
    `Docker <https://docs.docker.com/install/>`__

task_decorators:
    docker: airflow.providers.docker.operators.docker.docker_decorator
```

(and then we can use the existing `import_string` util.)

Reply:
That's better for registry too: https://registry.astronomer.io/providers/google :)

Reply:
Cool. In the (near) future we can optimize it further. I think we could easily push caching one level deeper (to dictionary values) and only import hook class when particular hook/operators is used. Should be rather easy. 

And I hope eventually with the modern UI will be able to completely change the approach and move most of the Hook custom behaviours to declarative way which won't require ANY code imports in Providers which would be great (thought I don't think it will be easy for extra links - that one might be not as easy to replace with declarative way).

Reply:
Yeah, Links are probably always going to need code loading, but those are relatively few-and-far between, and also could also be loaded on demand on first access